### PR TITLE
refactor: improve par

### LINKED ISSUE #4535: Two improvements to `par`
Given `par (exp1, exp2, exp3)` or `par exp1(exp2, exp3)` it would be useful if:

- We only create `n - 1` spawns/channels since we can reuse the current thread for one of the expressions (say the last).
- If `exp_i` is a variable or literal then there is no reason to spawn anything. For example, in `par (exp1, 42, w, exp4)` we should only spawn for `exp1` and `exp4` (and actually only for say `exp1` (or `exp4`) since we can reuse the current thread.

### PR DESCRIPTION
**Pull request by jaschdoc**
Avoids spawning threads for literals and variables and only spawns n - 1 thread so we use the current thread for evaluating an expression (instead of just waiting for other threads).
Closes #4535

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/phase/Lowering.scala [hunk 1]
```diff
@@ -1369,24 +1369,41 @@ object Lowering {
 
   /**
     * Returns a list of `GetChannel` expressions based on `symExps`.
+    *
+    * If the expression is either [[Expression.Var]] or a literal expression a [[Expression.GetChannel]] is **not** generated.
+    *
     */
   private def mkParWaits(symExps: List[(Symbol.VarSym, Expression)]): List[Expression] = {
     // Make wait expressions `<- ch, ..., <- chn`.
+    var first = true
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Lowering.scala:1378]: Trust the force, i mean, functional programming

Comment (jaschdoc) [reply main/src/ca/uwaterloo/flix/language/phase/Lowering.scala:1378]: Yeah this was just meant as an experiment

**Comment by jaschdoc** [comment]
Does not currently spawn a thread for `a()` in `testPar12`. I'll take a closer look some other day with fresh eyes.

```
def Test/Exp/Par.testPar12(_unit%175656: Unit, ): Bool & Pure = 
    let channel%175606%175674: Channel[Bool] = Channel 1i32;
  let _%175975: Unit = spawn Closure(Test/Exp/Par.testPar12%176347, [Closure(Test/Exp/Par.testPar12%176346, []), channel%175606%175674, ]);
  let a%176626: Bool = true;
  let b%176627: Bool = let x%176624: List[Int32] = List.range(0i32, 2000i32, ));
    let f%176625: List[Int32] -> Bool = let %0%176623: Int32 -> Bool = Closure(Test/Exp/Par.testPar12%176341, []);
      Closure(Test/Exp/Par.testPar12%176339, [%0%176623, ]);
    f%176625(x%176624, );
  Closure(Test/Exp/Par.testPar12%176336, [b%176627, a%176626, ])(<- channel%175606%175674, ) as Bool

def Test/Exp/Par.testPar12%176339(%0%176277: Int32 -> Bool, %1%175665: List[Int32], ): Bool & Pure = 
    List.forall%175666*(%0%176277, %1%175665, )

def Test/Exp/Par.testPar12%176341(x%175667: Int32, ): Bool & Pure = 
    x%175667 > -1i32

def Test/Exp/Par.testPar12%176345(x%175673: Int32, ): Bool & Pure = 
    x%175673 > -1i32

def Test/Exp/Par.testPar12%176343(%0%176278: Int32 -> Bool, %1%175672: List[Int32], ): Bool & Pure = 
    List.forall%175666*(%0%176278, %1%175672, )

def Test/Exp/Par.testPar12%176342(_unit%175662: Unit, ): Bool & Pure = 
    let %0%176702: Int32 -> Bool = Closure(Test/Exp/Par.testPar12%176341, []);
  let %1%176703: List[Int32] = List.range(0i32, 2000i32, ));
  List.forall%175666*(%0%176702, %1%176703, )

def Test/Exp/Par.testPar12%176344(%0%175671: Int32 -> Bool, ): List[Int32] -> Bool & Pure = 
    Closure(Test/Exp/Par.testPar12%176343, [%0%175671, ])

def Test/Exp/Par.testPar12%176346(_unit%175670: Unit, ): Bool & Pure = 
    let %0%176708: Int32 -> Bool = Closure(Test/Exp/Par.testPar12%176345, []);
  let %1%176709: List[Int32] = List.range(0i32, 2000i32, ));
  List.forall%175666*(%0%176708, %1%176709, )

def Test/Exp/Par.testPar12%176337(a%176274: Bool, b%175659: Bool, ): Bool -> Bool & Pure = 
    Closure(Test/Exp/Par.testPar12%176336, [b%175659, a%176274, ])

def Test/Exp/Par.testPar12%176340(%0%175664: Int32 -> Bool, ): List[Int32] -> Bool & Pure = 
    Closure(Test/Exp/Par.testPar12%176339, [%0%175664, ])

def Test/Exp/Par.testPar12%176338(a%175658: Bool, ): Bool -> (Bool -> Bool) & Pure = 
    Closure(Test/Exp/Par.testPar12%176337, [a%175658, ])

def Test/Exp/Par.testPar12%176347(b%176279: Unit -> Bool, channel%175606%176280: Channel[Bool], ): Unit & Impure = 
    channel%175606%176280 <- b%176279(Unit, )

def Test/Exp/Par.testPar12%176336(b%176275: Bool, a%176276: Bool, c%175660: Bool, ): Bool & Pure = 
    a%176276 and b%176275 and c%175660
```

**Comment by jaschdoc** [comment]
Here's the AST I get for `testPar13`:
```flix
    @test
    def testPar13(): Bool =
        let a = () -> List.range(0, 2000) |> List.forall(x -> x > -1);
        let b = () -> List.range(0, 2000) |> List.forall(x -> x > -1);
        let c = () -> List.range(0, 2000) |> List.forall(x -> x > -1);
        par h(a(), b(), c()) // Should spawn threads for `b()` and `c()`.
```
```

def Test/Exp/Par.testPar13%176655(%0%176563: Int32 -> Bool, %1%176083: List[Int32], ): Bool & Pure = 
    List.forall%175881*(%0%176563, %1%176083, )

def Test/Exp/Par.testPar13%176662(_unit%176086: Unit, ): Bool & Pure = 
    let %0%176912: Int32 -> Bool = Closure(Test/Exp/Par.testPar13%176661, []);
  let %1%176913: List[Int32] = List.range(0i32, 2000i32, ));
  List.forall%175881*(%0%176912, %1%176913, )

def Test/Exp/Par.testPar13%176654(_unit%176076: Unit, ): Bool & Pure = 
    let %0%176944: Int32 -> Bool = Closure(Test/Exp/Par.testPar13%176653, []);
  let %1%176945: List[Int32] = List.range(0i32, 2000i32, ));
  List.forall%175881*(%0%176944, %1%176945, )

def Test/Exp/Par.testPar13%176657(x%176084: Int32, ): Bool & Pure = 
    x%176084 > -1i32

def Test/Exp/Par.testPar13%176658(_unit%176081: Unit, ): Bool & Pure = 
    let %0%176956: Int32 -> Bool = Closure(Test/Exp/Par.testPar13%176657, []);
  let %1%176957: List[Int32] = List.range(0i32, 2000i32, ));
  List.forall%175881*(%0%176956, %1%176957, )

def Test/Exp/Par.testPar13%176663(b%176565: Unit -> Bool, channel%175700%176566: Channel[Bool], ): Unit & Impure = 
    channel%175700%176566 <- b%176565(Unit, )

def Test/Exp/Par.testPar13%176664(c%176567: Unit -> Bool, channel%175703%176568: Channel[Bool], ): Unit & Impure = 
    channel%175703%176568 <- c%176567(Unit, )

def Test/Exp/Par.testPar13%176660(%0%176087: Int32 -> Bool, ): List[Int32] -> Bool & Pure = 
    Closure(Test/Exp/Par.testPar13%176659, [%0%176087, ])

def Test/Exp/Par.testPar13%176653(x%176079: Int32, ): Bool & Pure = 
    x%176079 > -1i32

def Test/Exp/Par.testPar13%176656(%0%176082: Int32 -> Bool, ): List[Int32] -> Bool & Pure = 
    Closure(Test/Exp/Par.testPar13%176655, [%0%176082, ])

def Test/Exp/Par.testPar13(_unit%176074: Unit, ): Bool & Pure = 
    let channel%175700%176090: Channel[Bool] = Channel 1i32;
  let channel%175703%176091: Channel[Bool] = Channel 1i32;
  let _%176430: Unit = spawn Closure(Test/Exp/Par.testPar13%176663, [Closure(Test/Exp/Par.testPar13%176658, []), channel%175700%176090, ]);
  let _%176432: Unit = spawn Closure(Test/Exp/Par.testPar13%176664, [Closure(Test/Exp/Par.testPar13%176662, []), channel%175703%176091, ]);
  let y%176843: Bool = <- channel%175700%176090;
  let z%176844: Bool = <- channel%175703%176091;
  let x%176986: List[Int32] = List.range(0i32, 2000i32, ));
  let f%176987: List[Int32] -> Bool = let %0%176985: Int32 -> Bool = Closure(Test/Exp/Par.testPar13%176653, []);
    Closure(Test/Exp/Par.testPar13%176651, [%0%176985, ]);
  f%176987(x%176986, ) and y%176843 and z%176844 as Bool

def Test/Exp/Par.testPar13%176651(%0%176562: Int32 -> Bool, %1%176078: List[Int32], ): Bool & Pure = 
    List.forall%175881*(%0%176562, %1%176078, )

def Test/Exp/Par.testPar13%176652(%0%176077: Int32 -> Bool, ): List[Int32] -> Bool & Pure = 
    Closure(Test/Exp/Par.testPar13%176651, [%0%176077, ])

def Test/Exp/Par.testPar13%176661(x%176089: Int32, ): Bool & Pure = 
    x%176089 > -1i32

def Test/Exp/Par.testPar13%176659(%0%176564: Int32 -> Bool, %1%176088: List[Int32], ): Bool & Pure = 
    List.forall%175881*(%0%176564, %1%176088, )
```

**Comment by mlutze** [comment]
Here's what it looks like if I disable the optimizer:

```
def testPar13(_unit%174874: Unit, ): Bool & Pure = 
    let a%174875: Unit -> Bool = Closure(testPar13%175158, []);
  let b%174883: Unit -> Bool = Closure(testPar13%175162, []);
  let c%174888: Unit -> Bool = Closure(testPar13%175166, []);
  let channel%174844%174893: Channel[Bool] = Channel 1i32;
  let channel%174847%174894: Channel[Bool] = Channel 1i32;
  let _%175093: Unit = spawn Closure(testPar13%175167, [b%174883, channel%174844%174893, ]);
  let _%175095: Unit = spawn Closure(testPar13%175168, [c%174888, channel%174847%174894, ]);
  h(a%174875(Unit, ), <- channel%174844%174893, <- channel%174847%174894, )) as Bool
```

That looks more like what we wanted, right?

**Comment by magnus-madsen** [comment]
Depends on whether the types are correct?

**Comment by magnus-madsen** [comment]
I think this is biting of more than can be chewed. Instead of trying to fix both things at once, I would start much simpler.

For example:

```
par (1 + 1, 42)
```

could become

```
let ch1 = channel
let ch2 = channel
spawn (c1 <- 1 + 1)
c1 <- 42;
(<- c1, <- c2)
```

This would already been an improvement (even if it creates one more channel than needed).

**Comment by magnus-madsen** [comment]
(closing and then you can try a fresh start with the above :))

**Pull request closed without merging**